src/main/scala/com/fulcrumgenomics/bam/EstimateRnaSeqInsertSize.scala

```diff
@@ -0,0 +1,267 @@
+    val mateAlignmentEnd = CoordMath.getEnd(rec.getMateAlignmentStart, mateCigar.getReferenceLength)
+
+    // Get the overlapping genes
+    val leftMostBase     = Math.min(rec.getAlignmentStart, rec.getMateAlignmentStart)
+    val rightMostBase    = Math.max(rec.getAlignmentEnd, mateAlignmentEnd)
+    val recInterval      = new Interval(rec.getReferenceName, leftMostBase, rightMostBase)
+    val overlappingGenes = overlapDetector.getOverlaps(recInterval)
+
+    if (overlappingGenes.size() == 1 && overlappingGenes.forall(g => CoordMath.encloses(g.start, g.end, recInterval.getStart, recInterval.getEnd))) {
+      val recBlocks               = rec.getAlignmentBlocks.toList
+      val mateBlocks              = SAMUtils.getAlignmentBlocks(mateCigar, rec.getMateAlignmentStart, "mate cigar").toList
+      val mappedBases             = recBlocks.map(_.getLength).sum + mateBlocks.map(_.getLength).sum
+      val gene                    = overlappingGenes.head
+      val transcripts             = gene.iterator
+      var insertSize: Option[Int] = None
+      var sameInsertSize: Boolean = true
+      while (transcripts.hasNext && sameInsertSize) {
+        val transcript = transcripts.next()
+
+        // Ignore transcripts with too little overlap
+        val overlap    = numReadBasesOverlappingTranscript(recBlocks, transcript) + numReadBasesOverlappingTranscript(mateBlocks, transcript)
+        if (overlap / mappedBases.toDouble >= minimumOverlap) {
+          // get the 5' position of the record and its mate
+          val rec5Prime   = if (rec.getReadNegativeStrandFlag) rec.getAlignmentEnd else rec.getAlignmentStart
+          val mate5Prime  = if (rec.getMateNegativeStrandFlag) mateAlignmentEnd    else rec.getMateAlignmentStart
+
+          // calculate the distance in transcript space
+          val lower5Prime = Math.min(rec5Prime, mate5Prime)
+          val upper5Prime = Math.max(rec5Prime, mate5Prime)
+          val transcriptInsertSize = transcript.filter { exon => CoordMath.overlaps(lower5Prime, upper5Prime, exon.start, exon.end) }
+            .map { exon =>
+              CoordMath.getOverlap(lower5Prime, upper5Prime, exon.start, exon.end)
+            }
+            .sum
+
+          // Make sure that the insert size is the same as any previous result
+          insertSize match {
+            case None        => insertSize = Some(transcriptInsertSize)
+            case Some(isize) => sameInsertSize = isize == transcriptInsertSize
+          }
+        }
+      }
+
+      if (sameInsertSize) {
+        insertSize
+      }
+      else {
+        None
+      }
+    }
+    else {
+      None
+    }
+  }
+
+  /** Gets the number of read bases that overlap the exons. */
+  private def numReadBasesOverlappingTranscript(alignmentBlocks: List[AlignmentBlock], transcript: Transcript): Int = {
+    val blocks  = alignmentBlocks.iterator.bufferBetter
+    val exons   = transcript.iterator.bufferBetter
```

Comment:
There's an in build assumption here that the transcript always iterates exons in genomic order (as opposed to say transcript order).  I wonder if it wouldn't be better to not have an `iterator` method, but instead have two methods that provide access in genomic and transcript order and make it explicit, e.g. `transcript.exonsInGenomeOrder` and `transcript.exonsInTranscriptOrder` or similar.